Seeedmultichannel relay control

KoenBreeman opened the pull request:
# What does this implement/fix?

<!-- Quick description and explanation of changes -->

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#4122

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:
```yaml
# Example config.yaml
external_components:
  - source: github://pr#7233
    components: [seeedmultichannelrelay]
    refresh: 1h

seeedmultichannelrelay:
  id: SeedMultiChannelRelay_ID
  change_address_to: 0x01 # changes the address of the seeedmultichannelrelay (only runs once to update into the EPROM) then you have to change the i2c_address

switch:
  - platform: seeedmultichannelrelay
    id: relay_1
    name: relay1
    channel: 1
    seeedmultichannelrelay_id: SeedMultiChannelRelay_ID

  - platform: seeedmultichannelrelay
    id: relay_2
    name: relay2
    channel: 2
    seeedmultichannelrelay_id: SeedMultiChannelRelay_ID

  - platform: seeedmultichannelrelay
    id: relay_8
    name: relay8
    channel: 8
    seeedmultichannelrelay_id: SeedMultiChannelRelay_ID

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

KoenBreeman committed:
Created seeedmultichanelle relay

KoenBreeman committed:
Created seeedmultichanelle relay

KoenBreeman committed:
updated test to include i2c

KoenBreeman committed:
updated test to include i2c

KoenBreeman committed:
correct pins i2c for other esp32-c3 and 8266 and 2040

KoenBreeman committed:
correct pins i2c for other esp32-c3 and 8266 and 2040

KoenBreeman committed:
updated formatting of switch.cpp file according to test

KoenBreeman committed:
updated formatting of switch.cpp file according to test

KoenBreeman commented:
@max246, @jandechent, @iBobik
This is the new pull request, since the old one was full of errors.

KoenBreeman committed:
removed double reference to .h file

KoenBreeman committed:
removed double reference to .h file

KoenBreeman committed:
removed comments

KoenBreeman committed:
removed comments

KoenBreeman committed:
updated components and test according to remark jesserocks

KoenBreeman committed:
updated components and test according to remark jesserocks

KoenBreeman committed:
removed spaces at end of init.py

KoenBreeman committed:
removed spaces at end of init.py

KoenBreeman committed:
minor corrections

KoenBreeman committed:
...

KoenBreeman committed:
corrected merge errors

KoenBreeman committed:
...

KoenBreeman committed:
....

iBobik commented:
Hello, I’ve just upgrades esphome and moved to this PR. Device is M5STack AtomS3 and 4 channels relay.

I was able to compile it, but channels numbering has changed.

In PR 4684 I used channels 1, 2, 3, 4 but with this PR I experimentally found I have to use channels 2, 3, 4, 5.

KoenBreeman committed:
Updated write relay so that the corrected channels are written.

KoenBreeman commented:
> Hello, I’ve just upgrades esphome and moved to this PR. Device is M5STack AtomS3 and 4 channels relay.
> 
> I was able to compile it, but channels numbering has changed. In PR 4684 I used channels 1, 2, 3, 4 but with this PR I experimentally found I have to use channels 2, 3, 4, 5.
> 
> Anyway it work well.

Hi IBobik

I changed the write relay, I think now the channels 1 through 4 should be usable.

regards

Koen

KoenBreeman committed:
deleted space

KoenBreeman committed:
Merge branch 'dev' into Seeed-multichannel-relay-control

swoboda1337 commented:
Right now this PR is failing CI, that needs to be resolved before a review.

swoboda1337 committed:
Merge branch 'dev' into Seeed-multichannel-relay-control

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

swoboda1337 reviewed (changes requested):
CI is failing and tests need to be updated after the recent changes (take a look at existing i2c tests). Thanks

KoenBreeman committed:
Updated CODEOWNERS

KoenBreeman committed:
updated CODEOWNERS

iBobik commented:
Just updated to 2025.12.2 with this component, works.
> ```
> esp32:
>   board: esp32-s3-devkitc-1
>   variant: esp32s3
>   framework:
>     type: esp-idf
> ```

swoboda1337 commented on CODEOWNERS:
```diff
@@ -397,6 +397,7 @@ esphome/components/sdp3x/* @Azimath
 esphome/components/seeed_mr24hpc1/* @limengdu
 esphome/components/seeed_mr60bha2/* @limengdu
 esphome/components/seeed_mr60fda2/* @limengdu
+esphome/components/seeedmultichannelrelay/* @KoenBreeman
```
Other seeed components use underscores: `seeed_mr24hpc1`, `seeed_mr60bha2`, `seeed_mr60fda2`. This should probably be `seeed_multi_channel_relay` to match convention. Worth getting right before merge since its a public facing component name

swoboda1337 commented on esphome/components/seeedmultichannelrelay/seeedmultichannelrelay.h:
```diff
@@ -0,0 +1,78 @@
+static constexpr uint8_t CMD_READ_I2C_ADDR = 0x12;
+static constexpr uint8_t CMD_READ_FIRMWARE_VER = 0x13;
+
+enum class RelayBit : uint8_t {
+  RELAY1 = 1,
+  RELAY2 = 2,
+  RELAY3 = 3,
+  RELAY4 = 4,
+  RELAY5 = 5,
+  RELAY6 = 6,
+  RELAY7 = 7,
+  RELAY8 = 8
+};
```
This enum isnt referenced anywhere, remove it

swoboda1337 commented on esphome/components/seeedmultichannelrelay/seeedmultichannelrelay.h:
```diff
@@ -0,0 +1,78 @@
+ protected:
+  void write1_byte_(uint8_t register_address, uint8_t data);
+  uint8_t read1_byte_(uint8_t register_address);
+  uint8_t channel_state_;  // Value to save channel state
```
`channel_state_` isnt initialized. `turn_on_channel_` does `channel_state_ |= bit` so garbage bits could turn on random relays on first use. Should be `{0}` and ideally read the actual state back from the device in setup

swoboda1337 commented on esphome/components/seeedmultichannelrelay/seeedmultichannelrelay.h:
```diff
@@ -0,0 +1,78 @@
+  */
+  void turn_off_channel_(uint8_t channel);
+
+  void dump_config() override;
+
+  void setup() override;
```
`setup()` and `dump_config()` are public virtual overrides from Component, should be in the public section not protected

swoboda1337 commented on esphome/components/seeedmultichannelrelay/seeedmultichannelrelay.cpp:
```diff
@@ -0,0 +1,81 @@
+  return data;
+}
+
+/*! @brief Control the on/off of the specified relay.
+  *  @param number Bit number of relay (0~3).
+    @param state OFF = 0, ON = 1 . */
```
Comment says (0~3) but channels are 1 to 8

swoboda1337 commented on esphome/components/seeedmultichannelrelay/switch/__init__.py:
```diff
@@ -0,0 +1,42 @@
+
+DEPENDENCIES = ["seeedmultichannelrelay"]
+
+SeeedMultiChannelRelaySwitch = seeedmultichannelrelay_ns.class_(
+    "SeeedMultiChannelRelaySwitch", cg.Component, i2c.I2CDevice, switch.Switch
+)
```
Switch doesnt do I2C directly, it talks through the parent hub. Remove `i2c.I2CDevice` from the class definition. The C++ header correctly doesnt inherit it so this is just wrong

swoboda1337 commented on esphome/components/seeedmultichannelrelay/switch/__init__.py:
```diff
@@ -0,0 +1,42 @@
+    "SeeedMultiChannelRelaySwitch", cg.Component, i2c.I2CDevice, switch.Switch
+)
+
+CONF_INTERLOCK_WAIT_TIME = "interlock_wait_time"
+
+RelayBit_ = seeedmultichannelrelay_ns.enum("RelayBit", is_class=True)
```
`CONF_INTERLOCK_WAIT_TIME` and `RelayBit_` are defined but never used. The C++ side has full interlock logic but the config schema doesnt expose it. Either add the interlock config options like [gpio switch](https://github.com/esphome/esphome/blob/dev/esphome/components/gpio/switch/__init__.py) does or remove all the interlock code from both Python and C++

swoboda1337 commented on esphome/components/seeedmultichannelrelay/switch/seeedmultichannelrelay_switch.cpp:
```diff
@@ -0,0 +1,77 @@
+
+static const char *const TAG = "switch.SeeedMultiChannelRelay";
+
+float SeeedMultiChannelRelaySwitch::get_setup_priority() const { return setup_priority::HARDWARE; }
```
HARDWARE priority (800) runs before the parent hub which defaults to DATA (600). The switch setup calls `turn_on()`/`turn_off()` which writes to the relay through the parent, but the parent hasnt had its `setup()` called yet. If `change_address_to` is configured the address change wont have happened when this runs. Remove this override and let it default to DATA so the hub sets up first
